main.go

```diff
@@ -48,36 +50,68 @@ func main() {
 		return
 	}
 
-	err = validation.Validate(opts)
-	if err != nil {
-		logger.Printf("%s", err)
-		os.Exit(1)
+	if err = validation.Validate(opts); err != nil {
+		logger.Fatalf("%s", err)
 	}
 
 	validator := NewValidator(opts.EmailDomains, opts.AuthenticatedEmailsFile)
 	oauthproxy, err := NewOAuthProxy(opts, validator)
 	if err != nil {
-		logger.Errorf("ERROR: Failed to initialise OAuth2 Proxy: %v", err)
-		os.Exit(1)
+		logger.Fatalf("ERROR: Failed to initialise OAuth2 Proxy: %v", err)
 	}
 
 	rand.Seed(time.Now().UnixNano())
 
+	oauthProxyStop :=make(chan struct{}, 1)
+	metricsStop := startMetricsServer(opts.MetricsAddress, oauthProxyStop)
+
 	s := &Server{
 		Handler: oauthproxy,
 		Opts:    opts,
-		stop:    make(chan struct{}, 1),
+		stop:    oauthProxyStop,
 	}
 	// Observe signals in background goroutine.
 	go func() {
 		sigint := make(chan os.Signal, 1)
 		signal.Notify(sigint, os.Interrupt, syscall.SIGTERM)
 		<-sigint
-		s.stop <- struct{}{} // notify having caught signal
+		s.stop <- struct{}{} // notify having caught signal stop oauthproxy
+		close(metricsStop)   // and the metrics endpoint
 	}()
 	s.ListenAndServe()
 }
 
+// startMetricsServer will start the metrics server on the specified address.
+// It always return a channel to signal stop even when it does not run.
+func startMetricsServer(address string, oauthProxyStop chan struct{}) chan struct{} {
+	stop := make(chan struct{}, 1)
+
+	// Attempt to setup the metrics endpoint if we have an address
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
+		go func() {
+			// ListenAndServe always returns a non-nil error. After Shutdown or
+			// Close, the returned error is ErrServerClosed
+			if err := s.ListenAndServe(); err != http.ErrServerClosed {
```

Comment:
@JoelSpeed I've updated this to only log a message if we get an error that isn't `ErrServerClosed` and in that case, we go ahead and close the proxy stop channel as well.

My thinking here is that only the metrics server has encountered an error such as an invalid address, so we want to close that does more gracefully than calling `os.Exit()` via `logger.Fatal()`. I don't think we need to close the metrics channel as the main go routine closes and the process will stop but it sort of seems more correct/complete to do it anyway.

Reply:
> I don't think we need to close the metrics channel as the main go routine closes and the process will stop but it sort of seems more correct/complete to do it anyway.

I'm a bit confused, you say we don't need to but you are closing it below?

So if the start of the metrics server doesn't work, what happens to the main process, does it shutdown gracefully? (If you've tried this do you happen to have the log output?)